File: seed/generate.py
#!/usr/bin/env python3
"""Generates seed/workers.json and seed/transactions.json for the LAYAK demo."""
import json, random, sys, os
from datetime import datetime, timedelta, timezone
from collections import defaultdict

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "backend", "score"))
from score import ScoreInput, compute_score

random.seed(42)

DEMO_DATE = datetime(2026, 4, 26, tzinfo=timezone.utc)

# Six full/partial calendar months within the 6-month lookback window.
# SQL: occurred_at >= NOW() - INTERVAL '6 months' (= Oct 26 2025).
# We start Nov 1 so all months are clean full months.
MONTH_STARTS = [
    datetime(2025, 11,  1, tzinfo=timezone.utc),
    datetime(2025, 12,  1, tzinfo=timezone.utc),
    datetime(2026,  1,  1, tzinfo=timezone.utc),
    datetime(2026,  2,  1, tzinfo=timezone.utc),
    datetime(2026,  3,  1, tzinfo=timezone.utc),
    datetime(2026,  4,  1, tzinfo=timezone.utc),
]
MONTH_ENDS = MONTH_STARTS[1:] + [DEMO_DATE]

TXN_COUNTER = [0]


def rand_ts(start: datetime, end: datetime) -> datetime:
    delta = (end - start).total_seconds()
    for _ in range(100):
        t = start + timedelta(seconds=random.random() * delta)
        if t.weekday() != 6 and 8 <= t.hour <= 20:
            return t
    return start.replace(hour=10)


def generate_txns(worker_id: str, monthly_targets: list,
                  amt_lo: float, amt_hi: float, counterparties: list) -> list:
    txns = []
    avg_amt = (amt_lo + amt_hi) / 2
    for start, end, target in zip(MONTH_STARTS, MONTH_ENDS, monthly_targets):
        n = max(4, round(target / avg_amt) + random.randint(-2, 2))
        timestamps = sorted([rand_ts(start, end) for _ in range(n)])

        # Distribute amounts summing to target
        amounts = []
        remaining = float(target)
        for i in range(n - 1):
            lo = amt_lo
            hi = min(amt_hi, remaining - amt_lo * (n - 1 - i))
            hi = max(lo, hi)
            a = round(random.uniform(lo, hi), 2)
            amounts.append(a)
            remaining -= a
        amounts.append(round(max(amt_lo, min(amt_hi, remaining)), 2))
        random.shuffle(amounts)

        for ts, amt in zip(timestamps, amounts):
            TXN_COUNTER[0] += 1
            txns.append({
                "txn_id": f"txn_{worker_id}_{TXN_COUNTER[0]:05d}",
                "worker_id": worker_id,
                "amount_myr": float(round(amt, 2)),
                "counterparty_id": random.choice(counterparties),
                "channel": "tng_qr" if random.random() < 0.75 else "duitnow_qr",
                "occurred_at": ts.strftime("%Y-%m-%dT%H:%M:%SZ"),
            })
    return txns


workers = []
all_txns = []


def add_worker(wid, name, trade, zone, joined_at_str, monthly,
               amt_lo, amt_hi, n_customers, cust_prefix):
    counterparties = [f"{cust_prefix}_{i:04d}" for i in range(1, n_customers + 1)]
    workers.append({
        "worker_id": wid,
        "full_name": name,
        "trade": trade,
        "zone": zone,
        "tng_account_ref": f"tng_{wid}",
        "joined_at": joined_at_str,
        "status": "active",
    })
    txns = generate_txns(wid, monthly, amt_lo, amt_hi, counterparties)
    all_txns.extend(txns)


# ── HERO: Kumar (barber, Brickfields, target score ~729 → "excellent") ────────
# monthly [3800,3950,4100,4050,3900,4200], 55 customers, tenure 18mo
add_worker(
    "worker_kumar_001", "Kumar Selvarajan",
    "barber_services", "brickfields_kl",
    "2024-10-01T00:00:00Z",
    [3800, 3950, 4100, 4050, 3900, 4200],
    35, 60, 55, "cust_kumar",
)

# ── HERO: Siti (kuih seller, Cheras) ─────────────────────────────────────────
add_worker(
    "worker_siti_001", "Siti Rahimah",
    "kuih_seller", "cheras_kl",
    "2024-06-01T00:00:00Z",
    [2800, 2950, 3100, 3050, 2900, 3200],
    8, 25, 75, "cust_siti",
)

# ── Barbers: 28 more → 30 total ───────────────────────────────────────────────
BARBERS = [
    ("Ravi Kumar",     "brickfields_kl"), ("Sham Selvam",    "chow_kit_kl"),
    ("Muthu Krishnan", "setapak_kl"),     ("Dinesh Pillai",  "brickfields_kl"),
    ("Suresh Nair",    "chow_kit_kl"),    ("Azlan Hamid",    "setapak_kl"),
    ("Farid Zulkifli", "brickfields_kl"), ("Hafiz Rahman",   "chow_kit_kl"),
    ("Rizal Hasan",    "setapak_kl"),     ("Zamri Idris",    "brickfields_kl"),
    ("Ahmad Razif",    "chow_kit_kl"),    ("Kamarul Ariff",  "setapak_kl"),
    ("Shafiq Rosli",   "brickfields_kl"), ("Lokman Hakim",   "chow_kit_kl"),
    ("Azhar Malik",    "setapak_kl"),     ("Bala Krishnan",  "brickfields_kl"),
    ("Gopal Raj",      "chow_kit_kl"),    ("Venkat Raman",   "setapak_kl"),
    ("Senthil Kumar",  "brickfields_kl"), ("Prakash Pillai", "chow_kit_kl"),
    ("Izwan Kamal",    "setapak_kl"),     ("Shahril Azam",   "brickfields_kl"),
    ("Fadzli Nordin",  "chow_kit_kl"),    ("Hazli Jamil",    "setapak_kl"),
    ("Khairul Azri",   "brickfields_kl"), ("Syafiq Zuhri",   "chow_kit_kl"),
    ("Wan Khairul",    "setapak_kl"),     ("Mohd Razali",    "brickfields_kl"),
]
for i, (name, zone) in enumerate(BARBERS, 2):
    wid = f"worker_barber_{i:03d}"
    months_ago = random.randint(6, 36)
    joined = (DEMO_DATE - timedelta(days=months_ago * 30)).strftime("%Y-%m-%dT00:00:00Z")
    base = random.uniform(3200, 5800)
    monthly = [round(base * random.uniform(0.88, 1.12)) for _ in range(6)]
    add_worker(wid, name, "barber_services", zone, joined, monthly,
               35, 65, random.randint(35, 80), f"cust_b{i}")

# ── Kuih sellers: 11 more → 12 total (incl. Siti) ────────────────────────────
KUIH = [
    ("Aishah Yusof",  "cheras_kl"),      ("Rosnah Majid",  "ampang_kl"),
    ("Halimah Bakar", "wangsa_maju_kl"), ("Zainab Alias",  "cheras_kl"),
    ("Kamariah Daud", "ampang_kl"),       ("Rohani Hassan", "wangsa_maju_kl"),
    ("Noraini Talib", "cheras_kl"),       ("Maimunah Zain", "ampang_kl"),
    ("Fauziah Rahim", "wangsa_maju_kl"), ("Ramlah Ismail", "cheras_kl"),
    ("Rugayah Ahmad", "ampang_kl"),
]
for i, (name, zone) in enumerate(KUIH, 2):
    wid = f"worker_kuih_{i:03d}"
    months_ago = random.randint(6, 36)
    joined = (DEMO_DATE - timedelta(days=months_ago * 30)).strftime("%Y-%m-%dT00:00:00Z")
    base = random.uniform(2000, 4200)
    monthly = [round(base * random.uniform(0.85, 1.15)) for _ in range(6)]
    add_worker(wid, name, "kuih_seller", zone, joined, monthly,
               8, 25, random.randint(50, 100), f"cust_k{i}")

# ── Mechanics: 8 total ────────────────────────────────────────────────────────
MECHS = [
    ("Chong Wei Ming", "kepong_kl"),  ("Lim Boon Tat",  "setapak_kl"),
    ("Tan Ah Kow",     "cheras_kl"),  ("Wong Fook Meng", "kepong_kl"),
    ("Lee Swee Huat",  "setapak_kl"), ("Ng Beng Hock",  "cheras_kl"),
    ("Ong Ah Seng",    "kepong_kl"),  ("Yap Kim Leng",  "setapak_kl"),
]
for i, (name, zone) in enumerate(MECHS, 1):
    wid = f"worker_mech_{i:03d}"
    months_ago = random.randint(6, 36)
    joined = (DEMO_DATE - timedelta(days=months_ago * 30)).strftime("%Y-%m-%dT00:00:00Z")
    base = random.uniform(4500, 9000)
    monthly = [round(base * random.uniform(0.80, 1.20)) for _ in range(6)]
    add_worker(wid, name, "mobile_mechanic", zone, joined, monthly,
               80, 280, random.randint(30, 60), f"cust_m{i}")


# ── Write output ──────────────────────────────────────────────────────────────
out_dir = os.path.dirname(__file__)

with open(os.path.join(out_dir, "workers.json"), "w") as f:
    json.dump(workers, f, indent=2)

with open(os.path.join(out_dir, "transactions.json"), "w") as f:
    json.dump(all_txns, f, indent=2)

print(f"workers: {len(workers)}")
print(f"transactions: {len(all_txns)}")

# ── Verify Kumar's projected score ────────────────────────────────────────────
kumar_txns = [t for t in all_txns if t["worker_id"] == "worker_kumar_001"]
monthly_map = defaultdict(float)
counterparties = set()
last30 = 0
cutoff_6mo = DEMO_DATE - timedelta(days=180)
cutoff_30d  = DEMO_DATE - timedelta(days=30)

for t in kumar_txns:
    ts = datetime.strptime(t["occurred_at"], "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc)
    if ts >= cutoff_6mo:
        monthly_map[ts.strftime("%Y-%m")] += t["amount_myr"]
        counterparties.add(t["counterparty_id"])
    if ts >= cutoff_30d:
        last30 += 1

monthly_list = [monthly_map[k] for k in sorted(monthly_map)]
avg = sum(monthly_list) / len(monthly_list) if monthly_list else 0
tenure = max(1, int((DEMO_DATE - datetime(2024, 10, 1, tzinfo=timezone.utc)).days / 30))

inp = ScoreInput(
    monthly_earnings=monthly_list,
    tenure_months=tenure,
    unique_counterparties=len(counterparties),
    earnings_6mo_avg=avg,
    txns_last_30d=last30,
)
result = compute_score(inp)
print(f"\nKumar projected DB score: {result.score} ({result.tier})")
print(f"  monthly sums : {[round(m) for m in monthly_list]}")
print(f"  avg          : {round(avg)}")
print(f"  counterparties: {len(counterparties)}")
print(f"  last 30d txns: {last30}")
print(f"  tenure       : {tenure} months")

File: seed/generate.js
#!/usr/bin/env node
"use strict";
const fs   = require("fs");
const path = require("path");

// ── Seeded PRNG (Mulberry32) ──────────────────────────────────────────────────
let _seed = 42;
function rand() {
  _seed += 0x6D2B79F5;
  let t = _seed;
  t = Math.imul(t ^ (t >>> 15), t | 1);
  t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
  return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
}
function randInt(lo, hi)    { return Math.floor(rand() * (hi - lo + 1)) + lo; }
function randFloat(lo, hi)  { return rand() * (hi - lo) + lo; }
function randChoice(arr)    { return arr[Math.floor(rand() * arr.length)]; }
function shuffle(arr) {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(rand() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
}

// ── Date helpers ──────────────────────────────────────────────────────────────
const DEMO_DATE = new Date("2026-04-26T00:00:00Z");

const MONTH_STARTS = [
  new Date("2025-11-01T00:00:00Z"),
  new Date("2025-12-01T00:00:00Z"),
  new Date("2026-01-01T00:00:00Z"),
  new Date("2026-02-01T00:00:00Z"),
  new Date("2026-03-01T00:00:00Z"),
  new Date("2026-04-01T00:00:00Z"),
];
const MONTH_ENDS = [...MONTH_STARTS.slice(1), DEMO_DATE];

function fmtTs(d) {
  return d.toISOString().replace(".000Z", "Z").replace(/\.\d+Z$/, "Z");
}

function randTs(start, end) {
  const span = end - start;
  for (let attempts = 0; attempts < 200; attempts++) {
    const t = new Date(start.getTime() + rand() * span);
    const dow  = t.getUTCDay();  // 0=Sun
    const hour = t.getUTCHours();
    if (dow !== 0 && hour >= 8 && hour <= 20) return t;
  }
  return new Date(start.getTime() + 3600000 * 10);
}

// ── Score formula (mirrors backend/score/score.py) ────────────────────────────
function computeScore(inp) {
  const { monthlyEarnings, tenureMonths, uniqueCounterparties,
          earnings6moAvg, txnsLast30d } = inp;
  const n    = monthlyEarnings.length;
  const mean = monthlyEarnings.reduce((a, b) => a + b, 0) / n;
  const variance = monthlyEarnings.reduce((s, x) => s + (x - mean) ** 2, 0) / n;
  const std  = Math.sqrt(variance);

  const consistency = mean > 0 ? Math.max(0, 1 - std / mean) : 0;
  const tenure      = tenureMonths > 0 ? Math.log(tenureMonths + 1) / Math.log(24) : 0;
  const diversity   = Math.min(1.0, uniqueCounterparties / 30);
  const volume      = Math.min(1.0, earnings6moAvg / 30000);
  const recency     = Math.min(1.0, txnsLast30d / 28);

  const raw = 0.30 * consistency + 0.20 * tenure + 0.20 * diversity
            + 0.20 * volume      + 0.10 * recency;
  const score = Math.max(300, Math.min(850, 300 + Math.round(raw * 550)));
  const tier  = score >= 700 ? "excellent" : score >= 600 ? "good"
              : score >= 500 ? "fair" : "building";

  return { score, tier, raw: Math.round(raw * 10000) / 10000,
           consistency: Math.round(consistency * 10000) / 10000 };
}

// ── Transaction generator ─────────────────────────────────────────────────────
let txnCounter = 0;

function generateTxns(workerId, monthlyTargets, amtLo, amtHi, counterparties) {
  const avgAmt = (amtLo + amtHi) / 2;
  const txns = [];

  for (let m = 0; m < MONTH_STARTS.length; m++) {
    const start  = MONTH_STARTS[m];
    const end    = MONTH_ENDS[m];
    const target = monthlyTargets[m];
    const n      = Math.max(4, Math.round(target / avgAmt) + randInt(-2, 2));

    const timestamps = Array.from({ length: n }, () => randTs(start, end))
                            .sort((a, b) => a - b);

    // Distribute amounts summing close to target
    const amounts = [];
    let remaining = target;
    for (let i = 0; i < n - 1; i++) {
      const lo = amtLo;
      const hi = Math.max(lo, Math.min(amtHi, remaining - amtLo * (n - 1 - i)));
      const a  = Math.round(randFloat(lo, hi) * 100) / 100;
      amounts.push(a);
      remaining -= a;
    }
    amounts.push(Math.round(Math.max(amtLo, Math.min(amtHi, remaining)) * 100) / 100);
    shuffle(amounts);

    for (let i = 0; i < n; i++) {
      txnCounter++;
      txns.push({
        txn_id:           `txn_${workerId}_${String(txnCounter).padStart(5, "0")}`,
        worker_id:        workerId,
        amount_myr:       amounts[i],
        counterparty_id:  randChoice(counterparties),
        channel:          rand() < 0.75 ? "tng_qr" : "duitnow_qr",
        occurred_at:      fmtTs(timestamps[i]),
      });
    }
  }
  return txns;
}

// ── Worker / data definitions ─────────────────────────────────────────────────
const workers   = [];
const allTxns   = [];

function addWorker(wid, name, trade, zone, joinedAt, monthly, amtLo, amtHi, nCust, custPrefix) {
  const counterparties = Array.from({ length: nCust }, (_, i) =>
    `${custPrefix}_${String(i + 1).padStart(4, "0")}`);
  workers.push({ worker_id: wid, full_name: name, trade, zone,
                 tng_account_ref: `tng_${wid}`, joined_at: joinedAt, status: "active" });
  allTxns.push(...generateTxns(wid, monthly, amtLo, amtHi, counterparties));
}

// Heroes
addWorker("worker_kumar_001", "Kumar Selvarajan",
          "barber_services", "brickfields_kl", "2024-10-01T00:00:00Z",
          [3800, 3950, 4100, 4050, 3900, 4200], 35, 60, 55, "cust_kumar");

addWorker("worker_siti_001", "Siti Rahimah",
          "kuih_seller", "cheras_kl", "2024-06-01T00:00:00Z",
          [2800, 2950, 3100, 3050, 2900, 3200], 8, 25, 75, "cust_siti");

// 28 more barbers
const BARBERS = [
  ["Ravi Kumar",     "brickfields_kl"], ["Sham Selvam",    "chow_kit_kl"],
  ["Muthu Krishnan", "setapak_kl"],     ["Dinesh Pillai",  "brickfields_kl"],
  ["Suresh Nair",    "chow_kit_kl"],    ["Azlan Hamid",    "setapak_kl"],
  ["Farid Zulkifli", "brickfields_kl"], ["Hafiz Rahman",   "chow_kit_kl"],
  ["Rizal Hasan",    "setapak_kl"],     ["Zamri Idris",    "brickfields_kl"],
  ["Ahmad Razif",    "chow_kit_kl"],    ["Kamarul Ariff",  "setapak_kl"],
  ["Shafiq Rosli",   "brickfields_kl"], ["Lokman Hakim",   "chow_kit_kl"],
  ["Azhar Malik",    "setapak_kl"],     ["Bala Krishnan",  "brickfields_kl"],
  ["Gopal Raj",      "chow_kit_kl"],    ["Venkat Raman",   "setapak_kl"],
  ["Senthil Kumar",  "brickfields_kl"], ["Prakash Pillai", "chow_kit_kl"],
  ["Izwan Kamal",    "setapak_kl"],     ["Shahril Azam",   "brickfields_kl"],
  ["Fadzli Nordin",  "chow_kit_kl"],    ["Hazli Jamil",    "setapak_kl"],
  ["Khairul Azri",   "brickfields_kl"], ["Syafiq Zuhri",   "chow_kit_kl"],
  ["Wan Khairul",    "setapak_kl"],     ["Mohd Razali",    "brickfields_kl"],
  ["Saiful Bahari",  "chow_kit_kl"],
];
BARBERS.forEach(([name, zone], i) => {
  const wid       = `worker_barber_${String(i + 2).padStart(3, "0")}`;
  const mthsAgo   = randInt(6, 36);
  const joined    = new Date(DEMO_DATE - mthsAgo * 30 * 86400000)
                        .toISOString().replace(/T.*/, "T00:00:00Z");
  const base      = randFloat(3200, 5800);
  const monthly   = Array.from({ length: 6 }, () => Math.round(base * randFloat(0.88, 1.12)));
  addWorker(wid, name, "barber_services", zone, joined, monthly,
            35, 65, randInt(35, 80), `cust_b${i + 2}`);
});

// 11 more kuih sellers
const KUIH = [
  ["Aishah Yusof",  "cheras_kl"],      ["Rosnah Majid",  "ampang_kl"],
  ["Halimah Bakar", "wangsa_maju_kl"], ["Zainab Alias",  "cheras_kl"],
  ["Kamariah Daud", "ampang_kl"],       ["Rohani Hassan", "wangsa_maju_kl"],
  ["Noraini Talib", "cheras_kl"],       ["Maimunah Zain", "ampang_kl"],
  ["Fauziah Rahim", "wangsa_maju_kl"], ["Ramlah Ismail", "cheras_kl"],
  ["Rugayah Ahmad", "ampang_kl"],
];
KUIH.forEach(([name, zone], i) => {
  const wid     = `worker_kuih_${String(i + 2).padStart(3, "0")}`;
  const mthsAgo = randInt(6, 36);
  const joined  = new Date(DEMO_DATE - mthsAgo * 30 * 86400000)
                      .toISOString().replace(/T.*/, "T00:00:00Z");
  const base    = randFloat(2000, 4200);
  const monthly = Array.from({ length: 6 }, () => Math.round(base * randFloat(0.85, 1.15)));
  addWorker(wid, name, "kuih_seller", zone, joined, monthly,
            8, 25, randInt(50, 100), `cust_k${i + 2}`);
});

// 8 mechanics
const MECHS = [
  ["Chong Wei Ming", "kepong_kl"],  ["Lim Boon Tat",   "setapak_kl"],
  ["Tan Ah Kow",     "cheras_kl"],  ["Wong Fook Meng",  "kepong_kl"],
  ["Lee Swee Huat",  "setapak_kl"], ["Ng Beng Hock",   "cheras_kl"],
  ["Ong Ah Seng",    "kepong_kl"],  ["Yap Kim Leng",   "setapak_kl"],
];
MECHS.forEach(([name, zone], i) => {
  const wid     = `worker_mech_${String(i + 1).padStart(3, "0")}`;
  const mthsAgo = randInt(6, 36);
  const joined  = new Date(DEMO_DATE - mthsAgo * 30 * 86400000)
                      .toISOString().replace(/T.*/, "T00:00:00Z");
  const base    = randFloat(4500, 9000);
  const monthly = Array.from({ length: 6 }, () => Math.round(base * randFloat(0.80, 1.20)));
  addWorker(wid, name, "mobile_mechanic", zone, joined, monthly,
            80, 280, randInt(30, 60), `cust_m${i + 1}`);
});

// ── Write output ──────────────────────────────────────────────────────────────
const outDir = __dirname;
fs.writeFileSync(path.join(outDir, "workers.json"),      JSON.stringify(workers,  null, 2));
fs.writeFileSync(path.join(outDir, "transactions.json"), JSON.stringify(allTxns, null, 2));
console.log(`workers      : ${workers.length}`);
console.log(`transactions : ${allTxns.length}`);

// ── Verify Kumar's projected DB score ─────────────────────────────────────────
const cutoff6mo = new Date(DEMO_DATE - 180 * 86400000);
const cutoff30d = new Date(DEMO_DATE -  30 * 86400000);

const kumarTxns = allTxns.filter(t => t.worker_id === "worker_kumar_001");
const monthlyMap = {};
const counterparties = new Set();
let last30 = 0;

for (const t of kumarTxns) {
  const ts = new Date(t.occurred_at);
  if (ts >= cutoff6mo) {
    const mo = t.occurred_at.slice(0, 7);
    monthlyMap[mo] = (monthlyMap[mo] || 0) + t.amount_myr;
    counterparties.add(t.counterparty_id);
  }
  if (ts >= cutoff30d) last30++;
}

const monthlyList = Object.keys(monthlyMap).sort().map(k => monthlyMap[k]);
const avg6mo      = monthlyList.reduce((a, b) => a + b, 0) / monthlyList.length;
const joinedDate  = new Date("2024-10-01T00:00:00Z");
const tenure      = Math.max(1, Math.floor((DEMO_DATE - joinedDate) / (30 * 86400000)));

const result = computeScore({
  monthlyEarnings:      monthlyList,
  tenureMonths:         tenure,
  uniqueCounterparties: counterparties.size,
  earnings6moAvg:       avg6mo,
  txnsLast30d:          last30,
});

console.log(`\nKumar projected DB score: ${result.score} (${result.tier})`);
console.log(`  monthly sums  : [${monthlyList.map(x => Math.round(x)).join(", ")}]`);
console.log(`  avg 6mo       : ${Math.round(avg6mo)}`);
console.log(`  counterparties: ${counterparties.size}`);
console.log(`  last 30d txns : ${last30}`);
console.log(`  tenure        : ${tenure} months`);
console.log(`  consistency   : ${result.consistency}`);
console.log(`  raw           : ${result.raw}`);
